DocGen4/Output/References.lean

```diff
@@ -0,0 +1,77 @@
+import DocGen4.Output.Template
+
+/-!
+
+# Generic functions for references support
+
+This file contains functions for references support,
+independent of actual implementation.
+
+The main function is `DocGen4.preprocessBibFile` which preprocess
+the contents of bib file using user provided `process` function,
+and save the bib file and processed json file to output directory.
+
+Currently we have an implementation of `process` function using `pybtex`,
+which is `DocGen4.Pybtex.process`. In the future this may be replaced by
+a pure Lean implementation.
+
+-/
+
+open Lean DocGen4 Output
+
+namespace DocGen4
+
+/-- Preprocess (using the user provided `process` function)
+and save the bib file to the output path. -/
+def preprocessBibFile (contents : String) (process : String → IO (Array BibItem)) : IO Unit := do
+  -- create directories
+  IO.FS.createDirAll basePath
+  IO.FS.createDirAll declarationsBasePath
+  -- erase all files
+  IO.FS.writeFile (basePath / "references.bib") contents
+  IO.FS.writeFile (declarationsBasePath / "references.json") "[]"
```

Comment:
Is this just a micro optimisation to not run `process` if the content is empty?

Reply:
Yes. Also given the fact that currently `process` is implemented by calling external program `pybtex`, when the content is empty (which occurs if the user is not using references at all) this function will still work even if `pybtex` is not installed.